### PR TITLE
feat: Billing Service

### PR DESCRIPTION
**Pull request by exception**
## What does this PR do?

Introduces a new Billing Service to Platform API. For now, it is used to automate billing and usage for platform-enabled teams. 

Relies upon Stripe for handling subscriptions.

### How to Test

- You need to have a valid platform-enabled team locally
- You need to configure `config/stripe.config.json` for the plans you need to test
- You need Stripe API key and you need to install stripe client
- Run `stripe listen --events customer.subscription.created,customer.subscription.updated --forward-to localhost:5555/api/v2/billing/webhook`
- Do a `POST` to `localhost:5555/api/v2/billing/{{your team id}}/subscribe`. The body needs to have the desired plan, eg `"plan": "STARTER"`. You will get a redirect URI, which you need to visit` and run through the subscribe flow. This will trigger a webhook call which will update your team with the required billing data on Prisma.

### CODE REVIEW

**Comment by keithwillcode** [comment]
@exception csn you please add a description

File: apps/api/v2/src/modules/billing/billing.service.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+  constructor(
+    private readonly teamsRepository: OrganizationsRepository,
+    public readonly stripeService: StripeService,
+    private readonly billingRepository: BillingRepository,
+    private readonly configService: ConfigService<AppConfig>
+  ) {
+    this.webAppUrl = configService.get("app.baseUrl", { infer: true }) ?? "https://app.cal.com";
+    this.plansToPriceId = new Map<PlatformPlan, string>();
+    // for (const plan in Object.keys(PlatformPlan)) {
```
Comment (keithwillcode) [inline apps/api/v2/src/modules/billing/billing.service.ts:25]: Delete this?

File: apps/api/v2/src/modules/oauth-clients/oauth-client.repository.ts [hunk 2]
```diff
@@ -59,7 +59,7 @@ export class OAuthClientRepository {
   }
 
   async getOAuthClientWithRefreshSecret(clientId: string, clientSecret: string, refreshToken: string) {
-    return await this.dbRead.prisma.platformOAuthClient.findFirst({
+    return this.dbRead.prisma.platformOAuthClient.findFirst({
```
Comment (keithwillcode) [inline apps/api/v2/src/modules/oauth-clients/oauth-client.repository.ts:62]: Why no await here?

File: apps/api/v2/src/modules/stripe/stripe.service.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import { ConfigService } from "@nestjs/config";
+import Stripe from "stripe";
+
+@Injectable()
+export class StripeService {
+  public stripe: Stripe;
+
+  constructor(configService: ConfigService<AppConfig>) {
+    this.stripe = new Stripe(configService.get("stripe.apiKey", { infer: true }) ?? "", {
+      apiVersion: "2020-08-27",
```
Comment (keithwillcode) [inline apps/api/v2/src/modules/stripe/stripe.service.ts:12]: Why this version?

**Review by keithwillcode — COMMENTED** [review]
Overall looks good. 

Had some questions and we should add some tests.

File: apps/api/v2/src/modules/stripe/stripe.service.ts (hunk 3 above)
Comment (exception) [reply apps/api/v2/src/modules/stripe/stripe.service.ts:12]: No reason, just the one it auto filled, will set to latest.

File: apps/api/v2/src/modules/oauth-clients/oauth-client.repository.ts (hunk 2 above)
Comment (exception) [reply apps/api/v2/src/modules/oauth-clients/oauth-client.repository.ts:62]: It's already a Promise, it's being awaited in the actual usage.

File: apps/api/v2/src/modules/billing/controllers/billing.controller.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  @Get("/:teamId/check")
+  @UseGuards(NextAuthGuard, OrganizationRolesGuard)
+  @Roles(["OWNER", "ADMIN"])
+  async checkTeamBilling(@Param("teamId") teamId: number): Promise<CheckPlatformBillingResponseDto> {
+    const teamBilling = await this.billingService.getBillingData(teamId);
+
+    return {
+      status: "success",
+      valid: teamBilling.status === "valid",
+    };
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/billing/controllers/billing.controller.ts:52]: should we return the active plan of the user here ?

File: apps/api/v2/src/modules/billing/controllers/billing.controller.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    private readonly billingService: BillingService,
+    private readonly configService: ConfigService<AppConfig>
+  ) {
+    this.stripeWhSecret = configService.get("stripe.webhookSecret", { infer: true }) ?? "";
+  }
+
+  @Get("/:teamId/check")
+  @UseGuards(NextAuthGuard, OrganizationRolesGuard)
+  @Roles(["OWNER", "ADMIN"])
+  async checkTeamBilling(@Param("teamId") teamId: number): Promise<CheckPlatformBillingResponseDto> {
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/billing/controllers/billing.controller.ts:46]: should we use ApiResponse type here

File: apps/api/v2/src/modules/billing/controllers/billing.controller.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    };
+  }
+
+  @Post("/:teamId/subscribe")
+  @UseGuards(NextAuthGuard, OrganizationRolesGuard)
+  @Roles(["OWNER", "ADMIN"])
+  async subscribeTeamToStripe(
+    @Param("teamId") teamId: number,
+    @Body() input: SubscribeToPlanInput
+  ): Promise<SubscribeTeamToBillingResponseDto> {
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/billing/controllers/billing.controller.ts:61]: should we use ApiResponse type here

Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/billing/controllers/billing.controller.ts:61]: Promise<ApiResponse<SubscribeTeamToBillingResponseDto>>

**Comment by PeerRich** [comment]
can you record a loom.com demo0

File: apps/api/v2/src/modules/billing/controllers/billing.controller.ts (hunk 5 above)
Comment (exception) [reply apps/api/v2/src/modules/billing/controllers/billing.controller.ts:46]: What's the benefit against the Response Dto?

File: apps/api/v2/src/app.module.ts [hunk 7]
```diff
@@ -53,7 +55,17 @@ import { AppController } from "./app.controller";
 })
 export class AppModule implements NestModule {
   configure(consumer: MiddlewareConsumer): void {
-    consumer.apply(AppLoggerMiddleware).forRoutes("*");
-    consumer.apply(RewriterMiddleware).forRoutes("/");
+    consumer
+      .apply(RawBodyMiddleware)
```
Comment (exception) [inline apps/api/v2/src/app.module.ts:59]: Stripe requires a Buffer/String for the Webhooks to be built. Since NestJS by default parses everything as JSON, this was not letting the event through. That's why this change was required.

File: apps/api/v2/src/modules/billing/billing.service.ts [hunk 8]
```diff
@@ -63,17 +64,20 @@ export class BillingService {
         line_items: [
           {
             price: this.plansToPriceId.get(plan),
+            quantity: 1,
```
Comment (exception) [inline apps/api/v2/src/modules/billing/billing.service.ts:67]: This is having some sort of weird behaviour due to metered usage.

File: apps/api/v2/src/modules/stripe/stripe.service.ts (hunk 3 above)
Comment (exception) [reply apps/api/v2/src/modules/stripe/stripe.service.ts:12]: @keithwillcode this is actually hardcoded by Stripe to this version. Doesn't look like we can differ from it, all other Stripe usages in Web App use this version too, and it's set to default in our Stripe dashboard too.

File: apps/api/v2/src/middleware/app.logger.middleware.ts [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import { Injectable, NestMiddleware, Logger } from "@nestjs/common";
 import { Request, NextFunction } from "express";
 
-import { Response } from "@calcom/platform-types";
+import { Response } from "../../../../../packages/platform/types/dist";
```
[the remark this answers is not in the document]
Comment (exception) [reply apps/api/v2/src/middleware/app.logger.middleware.ts:4]: Huh, I am actually unsure why this changed...

File: apps/api/v2/src/modules/billing/controllers/billing.controller.ts (hunk 6 above)
Comment (exception) [reply apps/api/v2/src/modules/billing/controllers/billing.controller.ts:65]: Doesn't look like we're using ApiResponse much anymore, so I don't see a point on using it when we're already shaping responses as outputs as classes.

**Comment by keithwillcode** [comment]
Please don’t enable auto-merge.

File: apps/api/v2/src/modules/billing/billing.repository.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+          some: {
+            id: clientId,
+          },
+        },
+      },
+      select: {
+        id: true,
+      },
+    });
+  }
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/billing/billing.repository.ts:30]: this should probably be in the organization.repository ?

File: apps/api/v2/src/modules/billing/controllers/billing.controller.ts (hunk 6 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/billing/controllers/billing.controller.ts:64]: the main goal is to force a standard on the shape of api responses
 and so that that the DTO only targets the actual data sent back

Promise<ApiResponse < DTO> >

File: apps/api/v2/src/modules/oauth-clients/controllers/oauth-clients/oauth-clients.controller.ts [hunk 11]
```diff
@@ -60,11 +68,17 @@ export class OAuthClientsController {
   async createOAuthClient(
     @GetUser() user: UserWithProfile,
     @Body() body: CreateOAuthClientInput
-  ): Promise<CreateOAuthClientResponseDto> {
+  ): Promise<CreateOAuthClientResponseDto | CreateOauthClientRedirect> {
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/oauth-clients/controllers/oauth-clients/oauth-clients.controller.ts:71]: I think the Redirect part is not needed anymore ?

File: apps/web/lib/hooks/settings/organizations/platform/oauth-clients/usePersistOAuthClient.ts [hunk 12]
```diff
@@ -34,6 +39,8 @@ export const useCreateOAuthClient = (
     onSuccess: (data) => {
       if (data.status === SUCCESS_STATUS) {
         onSuccess?.();
+      } else if (data.status === REDIRECT_STATUS) {
+        window.location.href = data.url;
```
Comment (ThyMinimalDev) [inline apps/web/lib/hooks/settings/organizations/platform/oauth-clients/usePersistOAuthClient.ts:43]: seems like the api endpoint never sends redirect status ?

File: apps/api/v2/src/modules/billing/controllers/billing.controller.ts (hunk 6 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/modules/billing/controllers/billing.controller.ts:64]: apiResponse could be used directly in the dto as well I think, as you see fit

**Review by keithwillcode — APPROVED** [review]
